[XPU] support skip quant and xpu local quant

linwei210 opened the pull request:
support skip quant sensitive OP and use xpu local quant to stablize non-quant OP output.

linwei210 committed:
[XPU] support skip quant and xpu local quant

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/PaddlePaddle/Paddle-Lite?pullRequest=9399) <br/>All committers have signed the CLA.

zhupengyang commented on lite/backends/xpu/target_wrapper.cc:
```diff
@@ -169,6 +169,10 @@ LITE_THREAD_LOCAL std::shared_ptr<void> TargetWrapperXPU::xpu_stream_{nullptr};
     TargetWrapperXPU::multi_encoder_precision;  // NOLINT
 LITE_THREAD_LOCAL bool TargetWrapperXPU::multi_encoder_adaptive_seqlen{false};
+// local quant
+LITE_THREAD_LOCAL bool TargetWrapperXPU::xpu_local_quant{false};
```
TargetWrapperXPU 下的变量就不需要 xpu 前缀来标识了

linwei210 replied:
嗯，本来表示这是xpu特有的特性，可以去掉

linwei210 replied:
done

zhupengyang commented on lite/core/optimizer/mir/__xpu__static_kernel_pick_pass.cc:
```diff
@@ -319,6 +341,32 @@ void XPUStaticKernelPickPass::ForceUseInt8Kernel(
+    return;
+  }
+
+  static bool xpu_local_quant = GetBoolFromEnv("XPU_LOCAL_QUANT") ||
+                                lite::TargetWrapperXPU::xpu_local_quant;
```
为什么需要 static ？

linwei210 replied:
为了不要每次都去check，第一次check就可以

zhupengyang commented on lite/core/optimizer/mir/fusion/__xpu__multi_encoder_fuse_pass.cc:
```diff
@@ -603,81 +596,146 @@ class XPUSingleEncoderFuser : public FuseBase {
     }
+    // mul & matmul input/output max
+    op_desc.SetAttr<std::vector<float>>("fc_input_max", input_max);
+    op_desc.SetAttr<std::vector<std::vector<float>>>("Y0_max", weight_max);
```
不要放到 attrs 里面。建议用 tensor 存储，tensor 本来就包含维度信息

linwei210 replied:
attr支持tensor的vector 吗？   per channel里面每个max的长度会不一样，所以可能不是一个tensor，而是n个tensor

linwei210 committed:
rename xpu_local_quant to local_quant

newway commented on lite/kernels/xpu/__xpu__multi_encoder_compute.cc:
```diff
@@ -91,51 +96,45 @@ void XPUMultiEncoderCompute::prepare_quant_max(
+  VLOG(3) << "Total weight max value number: " << max_value_num;
+
+  if (!per_channel) {
+    max_value_num *= 6;
```
6可以换成max_ptr_len

linwei210 replied:
done

linwei210 committed:
use tensor to propagate weight max values

linwei210 committed:
use max_ptr_len instead of 6

zhupengyang commented on lite/backends/xpu/target_wrapper.cc:
```diff
@@ -169,6 +169,10 @@ LITE_THREAD_LOCAL std::shared_ptr<void> TargetWrapperXPU::xpu_stream_{nullptr};
 LITE_THREAD_LOCAL bool TargetWrapperXPU::multi_encoder_adaptive_seqlen{false};
+// local quant
+LITE_THREAD_LOCAL bool TargetWrapperXPU::local_quant{false};
+LITE_THREAD_LOCAL std::string
+    TargetWrapperXPU::xpu_compute_precision;  // NOLINT
```
这里也不需要 xpu_ 前缀标识

linwei210 replied:
done

zhupengyang commented on lite/core/optimizer/mir/fusion/__xpu__multi_encoder_fuse_pass.cc:
```diff
@@ -602,82 +599,160 @@ class XPUSingleEncoderFuser : public FuseBase {
+    std::vector<std::string> weight_max_tensor_name(quant_mul_ops.size());
+    CHECK(op_desc.HasInput("FCWeight")) << "op_desc does not have FCWeight Input.";
+    const auto& fc_weight_names = op_desc.Input("FCWeight");
+    CHECK(fc_weight_names.size() == quant_mul_ops.size()) << "FCWeight size is wrong.";
```
CHECK_EQ

linwei210 replied:
done

zhupengyang commented on lite/core/optimizer/mir/fusion/__xpu__multi_encoder_fuse_pass.cc:
```diff
@@ -902,44 +906,35 @@ class XPUMultiEncoderFuser {
-                << fc_weight_max[i] << ", " << fc_weight_max[i + 2];
-          }
-        }
+        // TODO : change to use tensor
+        // if (!per_channel) {
+        //   for (int i = 0; i < fc_weight_max.size(); i += 6) {
+        //     CHECK_LT(std::abs(fc_weight_max[i][0] - fc_weight_max[i + 1][0]), 1e-5)
+        //         << " quanted ernie's q/k weight scale should be euqal: "
+        //         << fc_weight_max[i][0] << ", " << fc_weight_max[i + 1][0];
+        //     CHECK_LT(std::abs(fc_weight_max[i][0] - fc_weight_max[i + 2][0]), 1e-5)
+        //         << " quanted ernie's q/v weight scale should be euqal: "
+        //         << fc_weight_max[i][0] << ", " << fc_weight_max[i + 2][0];
+        //   }
+        // }
```
不需要的代码就删掉

linwei210 replied:
done

zhupengyang commented on lite/core/optimizer/mir/fusion/__xpu__multi_encoder_fuse_pass.cc:
```diff
@@ -966,26 +958,22 @@ class XPUMultiEncoderFuser {
+        auto tag_tensor = scope->FindMutableTensor(update_tag);
+        if (tag_tensor != nullptr) {
+          auto max_tensor = scope->FindTensor(max_tensor_name);
+          CHECK(max_tensor != nullptr && max_tensor->numel() == 1);
```
分开 check，报错的时候比较容易知道是哪个条件有问题

linwei210 replied:
done

zhupengyang commented on lite/kernels/xpu/__xpu__fc_compute.cc:
```diff
@@ -267,3 +270,13 @@ REGISTER_LITE_KERNEL(
+    .BindInput("Bias", {LiteType::GetTensorTy(TARGET(kXPU))})
+    .BindOutput("Output", {LiteType::GetTensorTy(TARGET(kXPU))})
+    .BindOutput("OutputMax", {LiteType::GetTensorTy(TARGET(kXPU))})
+    .Finalize();
```
文件末尾加一个空行

linwei210 replied:
done

zhupengyang commented on lite/kernels/xpu/__xpu__multi_encoder_compute.cc:
```diff
@@ -259,6 +279,8 @@ void XPUMultiEncoderCompute::run_encoder(const T* in, T* out) {
 }
 
 void XPUMultiEncoderCompute::Run() {
+  static bool xpu_local_quant = GetBoolFromEnv("XPU_LOCAL_QUANT") ||
+                                lite::TargetWrapperXPU::local_quant;
```
只在第一次运行时执行的可以放在 PrepareForRun 里面，不需要用 static

linwei210 replied:
done

linwei210 committed:
fixed review comments

zhupengyang commented on lite/core/optimizer/mir/__xpu__static_kernel_pick_pass.cc:
```diff
@@ -319,6 +341,32 @@ void XPUStaticKernelPickPass::ForceUseInt8Kernel(
+    return;
+  }
+
+  static bool xpu_local_quant = GetBoolFromEnv("XPU_LOCAL_QUANT") ||
+                                lite::TargetWrapperXPU::local_quant;
```
这里也不要用static

pass 只在 create predictor 的时候跑一次，没有非常影响性能

linwei210 replied:
done

zhupengyang commented on lite/kernels/xpu/__xpu__multi_encoder_compute.cc:
```diff
@@ -337,6 +369,5 @@ REGISTER_LITE_KERNEL(__xpu__multi_encoder,
-    .BindInput("FCWeightMax", {LiteType::GetTensorTy(TARGET(kHost))})
     .BindOutput("Output", {LiteType::GetTensorTy(TARGET(kXPU))})
-    .Finalize();
+    .Finalize();
```
类似的问题都改下吧

linwei210 replied:
done

zhupengyang commented on lite/kernels/xpu/__xpu__multi_encoder_compute.h:
```diff
@@ -37,17 +37,20 @@ class XPUMultiEncoderCompute
   XPUScratchPadGuard cast_out_guard_;
   xdnn::Activation_t qkv_act = xdnn::Activation_t::RELU;
   int slice_idx = -1;
+  bool xpu_local_quant_ = false;
```
算子本身有 XPU 前缀，也不会和其它硬件混淆，看起来不需要用 xpu_ 前缀去标识这个变量

linwei210 replied:
done

linwei210 committed:
fixed format issue

zhupengyang reviewed (dismissed):
LGTM

linwei210 committed:
fixed AndroidNNAPI test error

non-const lvalue reference to type 'basic_string<...>' cannot bind to a temporary of type 'basic_string<...>'

linwei210 committed:
fixed clang-format

linwei210 committed:
fixed compilation error of non-xpu platform

linwei210 committed:
fixed other platform compilation error

linwei210 committed:
fixed compilation error

linwei210 committed:
fixed

linwei210 committed:
fixed

linwei210 committed:
fixed

linwei210 committed:
rebase

zhupengyang commented on lite/backends/xpu/math.h:
```diff
@@ -130,6 +130,16 @@ static inline int ConvertFP32ToInt8(const void *input,
   return 0;
 }
 
+#ifdef LITE_WITH_XPU
```
不需要加宏

zhupengyang commented on lite/api/paddle_use_passes.h:
```diff
@@ -87,6 +87,7 @@ USE_MIR_PASS(lite_flatten_fc_fuse_pass);
 USE_MIR_PASS(lite_fc_prelu_fuse_pass);
 USE_MIR_PASS(lite_greater_than_cast_fuse_pass);
 USE_MIR_PASS(assign_value_calc_offline_pass);
+#ifdef LITE_WITH_XPU
```
不需要加宏

zhupengyang commented on lite/core/optimizer/optimizer.cc:
```diff
@@ -206,6 +206,7 @@ std::unique_ptr<RuntimeProgram> RunDefaultOptimizer(
        "identity_dropout_eliminate_pass",
        "sparse_conv_detect_pass",
        "keepdims_convert_pass",
+#ifdef LITE_WITH_XPU
```
不需要加宏

zhupengyang commented on lite/core/optimizer/optimizer.cc:
```diff
@@ -223,9 +224,12 @@ std::unique_ptr<RuntimeProgram> RunDefaultOptimizer(
        "__xpu__link_previous_out_max_pass",
+#endif
        "fix_mismatched_precision_pass",
+#ifdef LITE_WITH_XPU
```
不需要加宏

linwei210 committed:
delete useless LITE_WITH_XPU marco

zhupengyang reviewed (approved):
LGTM
